File: src/main/java/com/company/erp/features/auth/service/AuthService.java
package com.company.erp.features.auth.service;

import com.company.erp.core.exceptions.ConflictException;
import com.company.erp.core.exceptions.NotFoundException;
import com.company.erp.core.exceptions.UnauthorizedException;
import com.company.erp.core.security.JwtService;
import com.company.erp.core.security.JwtService.IssuedToken;
import com.company.erp.core.security.JwtService.RefreshClaims;
import com.company.erp.features.auth.dto.*;
import com.company.erp.features.auth.entity.RefreshToken;
import com.company.erp.features.auth.repository.RefreshTokenRepository;
import com.company.erp.features.employees.service.EmployeeService;
import com.company.erp.features.users.dto.UserDto;
import com.company.erp.features.users.entity.User;
import com.company.erp.features.users.repository.UserRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;

@Service
@Transactional
public class AuthService {

    private final UserRepository users;
    private final RefreshTokenRepository refreshTokens;
    private final PasswordEncoder passwords;
    private final JwtService jwt;
    private final EmployeeService employees;

    public AuthService(UserRepository users,
                       RefreshTokenRepository refreshTokens,
                       PasswordEncoder passwords,
                       JwtService jwt,
                       EmployeeService employees) {
        this.users = users;
        this.refreshTokens = refreshTokens;
        this.passwords = passwords;
        this.jwt = jwt;
        this.employees = employees;
    }

    public AuthResponse login(LoginRequest req) {
        User user = users.findByEmail(req.email().toLowerCase())
                .orElseThrow(() -> new UnauthorizedException("Invalid email or password"));
        if (!user.isEnabled()) throw new UnauthorizedException("Account is disabled");
        if (!passwords.matches(req.password(), user.getPasswordHash())) {
            throw new UnauthorizedException("Invalid email or password");
        }
        employees.touchLastLoginByUserId(user.getId());
        return issueTokens(user);
    }

    public AuthResponse register(RegisterRequest req) {
        String email = req.email().toLowerCase();
        if (users.existsByEmail(email)) throw new ConflictException("Email already in use");

        User user = new User();
        user.setEmail(email);
        user.setPasswordHash(passwords.encode(req.password()));
        user.setFullName(req.fullName());
        user.setPhone(req.phone());
        users.save(user);
        return issueTokens(user);
    }

    public AuthResponse refresh(RefreshRequest req) {
        RefreshClaims claims = jwt.parseRefresh(req.refreshToken());
        RefreshToken stored = refreshTokens.findByJti(claims.jti())
                .orElseThrow(() -> new UnauthorizedException("Refresh token not recognised"));
        if (!stored.isActive(Instant.now())) throw new UnauthorizedException("Refresh token expired or revoked");
        if (!stored.getUserId().equals(claims.userId())) throw new UnauthorizedException("Refresh token user mismatch");

        // Rotate: revoke the presented token, issue a fresh pair.
        stored.setRevokedAt(Instant.now());

        User user = users.findWithRolesById(claims.userId())
                .orElseThrow(() -> new NotFoundException("User no longer exists"));
        return issueTokens(user);
    }

    public void logout(LogoutRequest req) {
        // Silent on invalid/unknown tokens — logout must always succeed from the client's POV.
        RefreshClaims claims;
        try {
            claims = jwt.parseRefresh(req.refreshToken());
        } catch (UnauthorizedException ex) {
            return;
        }
        refreshTokens.findByJti(claims.jti()).ifPresent(rt -> rt.setRevokedAt(Instant.now()));
    }

    private AuthResponse issueTokens(User user) {
        IssuedToken access  = jwt.issueAccess(user.getId(), user.getEmail(), user.allPermissions());
        IssuedToken refresh = jwt.issueRefresh(user.getId());

        RefreshToken record = new RefreshToken();
        record.setUserId(user.getId());
        record.setJti(refresh.jti());
        record.setIssuedAt(Instant.now());
        record.setExpiresAt(refresh.expiresAt());
        refreshTokens.save(record);

        return new AuthResponse(
                access.value(), access.expiresAt(),
                refresh.value(), refresh.expiresAt(),
                UserDto.from(user));
    }
}

File: app/Features/Auth/Services/AuthService.php
<?php

namespace App\Features\Auth\Services;

use App\Features\Auth\Dto\AuthResponse;
use App\Features\Auth\Models\RefreshToken;
use App\Features\Employees\Services\EmployeeService;
use App\Features\Users\Dto\UserDto;
use App\Features\Users\Models\User;
use App\Support\Auth\JwtService;
use App\Support\Exceptions\ConflictException;
use App\Support\Exceptions\NotFoundException;
use App\Support\Exceptions\UnauthorizedException;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;

class AuthService
{
    /** @var JwtService */
    private $jwt;

    /** @var EmployeeService */
    private $employees;

    public function __construct(JwtService $jwt, EmployeeService $employees)
    {
        $this->jwt = $jwt;
        $this->employees = $employees;
    }

    /**
     * @return array<string, mixed>
     */
    public function login(string $email, string $password): array
    {
        $user = User::query()->with('roles.permissions')
            ->where('email', strtolower($email))->first();

        if (! $user) {
            throw new UnauthorizedException('Invalid email or password');
        }
        if (! $user->enabled) {
            throw new UnauthorizedException('Account is disabled');
        }
        if (! Hash::check($password, $user->password_hash)) {
            throw new UnauthorizedException('Invalid email or password');
        }

        $this->employees->touchLastLoginByUserId((int) $user->id);

        return $this->issueTokens($user);
    }

    /**
     * @return array<string, mixed>
     */
    public function register(string $email, string $password, string $fullName, ?string $phone): array
    {
        $email = strtolower($email);
        if (User::query()->where('email', $email)->exists()) {
            throw new ConflictException('Email already in use');
        }

        $user = new User();
        $user->email = $email;
        $user->password_hash = Hash::make($password);
        $user->full_name = $fullName;
        $user->phone = $phone;
        $user->save();

        return $this->issueTokens($user->load('roles.permissions'));
    }

    /**
     * @return array<string, mixed>
     */
    public function refresh(string $refreshToken): array
    {
        $claims = $this->jwt->parseRefresh($refreshToken);

        return DB::transaction(function () use ($claims) {
            $stored = RefreshToken::query()->where('jti', $claims['jti'])->first();
            if (! $stored) {
                throw new UnauthorizedException('Refresh token not recognised');
            }
            if (! $stored->isActive(Carbon::now())) {
                throw new UnauthorizedException('Refresh token expired or revoked');
            }
            if ((int) $stored->user_id !== (int) $claims['userId']) {
                throw new UnauthorizedException('Refresh token user mismatch');
            }

            // Rotate: revoke the presented token, issue a fresh pair.
            $stored->revoked_at = Carbon::now();
            $stored->save();

            $user = User::query()->with('roles.permissions')->find($claims['userId']);
            if (! $user) {
                throw new NotFoundException('User no longer exists');
            }

            return $this->issueTokens($user);
        });
    }

    public function logout(string $refreshToken): void
    {
        // Silent on invalid/unknown tokens — logout must always succeed.
        try {
            $claims = $this->jwt->parseRefresh($refreshToken);
        } catch (UnauthorizedException $e) {
            return;
        }

        $stored = RefreshToken::query()->where('jti', $claims['jti'])->first();
        if ($stored) {
            $stored->revoked_at = Carbon::now();
            $stored->save();
        }
    }

    /**
     * @return array<string, mixed>
     */
    private function issueTokens(User $user): array
    {
        $access = $this->jwt->issueAccess((int) $user->id, $user->email, $user->allPermissions());
        $refresh = $this->jwt->issueRefresh((int) $user->id);

        $record = new RefreshToken();
        $record->user_id = (int) $user->id;
        $record->jti = $refresh['jti'];
        $record->issued_at = Carbon::now();
        $record->expires_at = Carbon::createFromTimestampUTC($refresh['expiresAt']);
        $record->save();

        return AuthResponse::build(
            $access['value'],
            $access['expiresAt'],
            $refresh['value'],
            $refresh['expiresAt'],
            UserDto::from($user)
        );
    }
}
